### PR TITLE
[READY] Remove Boost.Regex in favour of std variant

### PR DESCRIPTION
**Pull request by bstaletic**
The final piece in boost to std transition. Continuation of #705.

To make this work on linux with gcc `ReadUtf8File()` had to be changed to return a string vector. Otherwise no matches were ever found. Seems like a gcc bug.

That being said tests on linux are stll failing. On OSX on the other hand tests are all passing. The tests that are failing are:

- IdentifierCompleterTest.TagsEndToEndWorks
- IdentifierUtilsTest.ExtractIdentifiersFromTagsFileWorks

Even though the tests are failing on linux, reading the `basic.tags` file results in the expected behaviour (i.e. tags are provided for completion).

Note:
This leaves only boost::filesystem and boost::python. So `update_boost.py` should be updated, as well as some parts of `CMakeLists.txt` that mention boost threads that are no longer in use.

---
This change is [Reviewable](https://reviewable.io/reviews/valloric/ycmd/719)

### CODE REVIEW

**Comment by Valloric** [comment]
> To make this work on linux with gcc ReadUtf8File() had to be changed to return a string vector. 

Not bothered by that at all. :)

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[cpp/ycm/Utils.cpp, line 41 at r1](https://reviewable.io:443/reviews/valloric/ycmd/719#-Kf2QdcIvPfo-psi8Z5a:-Kf2QdcIvPfo-psi8Z5b:b91xtej) ([raw file](https://github.com/valloric/ycmd/blob/dee1e0008cfa21e94d6cb472fab5bf76068a9603/cpp/ycm/Utils.cpp#L41)):*
> ```cpp
>   while( getline( file, line ) )
>   {
> 	  contents.push_back(line);
> ```

Formatting looks off; seems to have a tab instead of spaces and the brace is in the wrong position.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/719#-:-Kf2QsecQwZTWsJox2Rz:b-abohcx)*

**Comment by bstaletic** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[cpp/ycm/Utils.cpp, line 41 at r1](https://reviewable.io:443/reviews/valloric/ycmd/719#-Kf2QdcIvPfo-psi8Z5a:-Kf2_3T9KPzQiJ5aArlr:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/dee1e0008cfa21e94d6cb472fab5bf76068a9603/cpp/ycm/Utils.cpp#L41)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Formatting looks off; seems to have a tab instead of spaces and the brace is in the wrong position.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/719)*

**Comment by Valloric** [comment]
There seem to be some more formatting issues; could you run our `style_format.sh` script from the repo root? It should take care of issues.

**Comment by bstaletic** [comment]
The script didn't do absolutely anything. So I did my best to manually fix all the styling issues.

As for the linux tests failing, I'm really confused that those fail only on Linux. Does anyone have a clue as to what is going on?

**Comment by vheon** [comment]
Reviewed 1 of 3 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cpp/ycm/Utils.cpp, line 43 at r3](https://reviewable.io:443/reviews/valloric/ycmd/719#-Kf8Qjn3lEqK51imbB-b:-Kf8Qjn42jhJUajiea7x:b-i5isxd) ([raw file](https://github.com/valloric/ycmd/blob/54ebfe848dcbbd000b0e4ae0a7fabf880be32fd2/cpp/ycm/Utils.cpp#L43)):*
> ```cpp
>   }
> 
>   if ( contents.size() == 0 )
> ```

Can't we just return `contents` here as well? if `contents.size()` is equal to `0`, returning it would be the same as `std::vector< std::string >()`, or am I missing something?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/719)*

**Comment by vheon** [comment]
From the error looks like there is some problem with the regex definition 😕  What I'm worried about is that the Travis build uses GCC 4.8.5 where the C++11 support is experimental apparently, so I'm not sure if is a real problem with the code right now or if it is the GCC regex library. I don't have my VM at hand now to check it :/ @bstaletic did you try our Vagrant env?

**Comment by bstaletic** [comment]
It fails on Vagrant env as well. I didn't run the tests on my host, which is Arch with GCC-6.3.1. The reason I didn't try it on host is because, when I quickly tried to set it up, the script `run_tests.py` failed to run.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cpp/ycm/Utils.cpp, line 43 at r3](https://reviewable.io:443/reviews/valloric/ycmd/719#-Kf8Qjn3lEqK51imbB-b:-Kf8rgQsvCCz7r11mUzZ:b-okww8h) ([raw file](https://github.com/valloric/ycmd/blob/54ebfe848dcbbd000b0e4ae0a7fabf880be32fd2/cpp/ycm/Utils.cpp#L43)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

Can't we just return `contents` here as well? if `contents.size()` is equal to `0`, returning it would be the same as `std::vector< std::string >()`, or am I missing something?
</blockquote></details>

You're right. No need for additional check.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/719#-:-Kf8pVdMF86zHu0feiff:b-y95hk8)*

**Comment by bstaletic** [comment]
Update:
I realised I could follow the full instalation guide and simply target `ycm_core_tests` instead of `ycm_core`. I wish I could say otherwise, but the tests failed on GCC-6.3.1 too.

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/719#-:-Kf9KFoYgMfdjBjWw2iH:bw3hi60)*

**Comment by bstaletic** [comment]
I have some more (good|bad)? news.

I've been playing with ycmd on another branch (my toy branch - labeled move fast and break things). When I had ran the core tests I noticed that all 68 of them are passing. That made me scratch my head. 

I went back to this branch and... again all tests are passing. Made some more tests trying to find the relevant change. Libclang? Bundled boost vs system provided? Since I made no changes to this branch neither made any sense.

So I played with different parameters for `build.py`. Yet whatever I had tried I couldn't get these tests to fail.

TL;DR:
On my machine the tests are passing. At least with GCC-6.3.1. No changes have been made to the branch.

I'll go and try Vagrant now, but it will take a while, as I need to reinitialise the environment.

**Comment by bstaletic** [comment]
First test on Vagrant is done.

With everything set to the default value in Vagrant env (python2.6, donwloading libclang and compiling bundled boost) it fails by throwing "bad regex" exception.

I wanted to test Vagrant on python3.3 (though I don't see python as a possible reason for test failures), but running `./build.py` produced:
```
CMake Error at /usr/share/cmake-2.8/Modules/FindPackageHandleStandardArgs.cmake:108 (message):
  Could NOT find PythonLibs (missing: PYTHON_INCLUDE_DIRS) (found suitable
  version "3.3.0", minimum required is "3.3")
```
after running `pyenv global 3.3.0`. I'll try to make python3 on Vagrant work and see what happens.

**Comment by bstaletic** [comment]
Since this still has a chance of getting merged, I solved the merge conflict introduced after merging #726.

**Comment by Bu11etmagnet** [comment]
Not the best of titles. There will be a `std::variant` in C++17 which is utterly unrelated to any of the regex classes.

**Comment by vheon** [comment]
@bstaletic didn't you reverted the change of #726?

**Comment by bstaletic** [comment]
@vheon Thanks for pointing that out.

Note to self: Don't rebase at 04:00AM.

File: cpp/ycm/IdentifierUtils.cpp [hunk 1]
```diff
@@ -114,38 +113,34 @@ const char *const NOT_FOUND = "YCMFOOBAR_NOT_FOUND";
...
-    std::string identifier( matches[ 1 ] );
-    fs::path path( matches[ 2 ].str() );
-    path = fs::absolute( path, path_to_tag_file.parent_path() )
-           .make_preferred();
-
-    filetype_identifier_map[ filetype ][ path.string() ].push_back( identifier );
+  std::smatch matches;
+  const std::regex expression( TAG_REGEX );
+
+  for ( auto line : tags_file_contents ) {
```
Comment (fsimonis) [inline cpp/ycm/IdentifierUtils.cpp]: Use `const auto& line` to prevent copy construction of lines.

File: cpp/ycm/IdentifierUtils.cpp [hunk 2]
```diff
@@ -37,9 +36,9 @@ const char *const TAG_REGEX =
   // The second field is the path to the file that has the identifier; either
   // absolute or relative to the tags file.
   "([^\\t\\n\\r]+)"
```
Comment (fsimonis) [inline cpp/ycm/IdentifierUtils.cpp]: You changed `ReadUtf8File` to return a vector of lines, thus you do not have to check for `\r` and `\n` anymore.

File: cpp/ycm/IdentifierUtils.cpp [hunk 3]
```diff
@@ -114,38 +113,34 @@ const char *const NOT_FOUND = "YCMFOOBAR_NOT_FOUND";
...
-      continue;
-
-    std::string identifier( matches[ 1 ] );
-    fs::path path( matches[ 2 ].str() );
-    path = fs::absolute( path, path_to_tag_file.parent_path() )
-           .make_preferred();
-
-    filetype_identifier_map[ filetype ][ path.string() ].push_back( identifier );
+  std::smatch matches;
+  const std::regex expression( TAG_REGEX );
```
Comment (fsimonis) [inline cpp/ycm/IdentifierUtils.cpp]: You can pass `std::regex::optimize` to the regex. It might be worth a try as the regex is used multiple times. A one time construction would be ideal of course, especially as `TAG_REGEX`is a `const`.

From [cppreference](http://en.cppreference.com/w/cpp/regex/syntax_option_type)
> Instructs the regular expression engine to make matching faster, with the potential cost of making construction slower. For example, this might mean converting a non-deterministic FSA to a deterministic FSA.

File: cpp/ycm/IdentifierUtils.cpp [hunk 4]
```diff
@@ -114,38 +113,34 @@ const char *const NOT_FOUND = "YCMFOOBAR_NOT_FOUND";
 FiletypeIdentifierMap ExtractIdentifiersFromTagsFile(
   const fs::path &path_to_tag_file ) {
   FiletypeIdentifierMap filetype_identifier_map;
-  std::string tags_file_contents;
+  std::vector< std::string > tags_file_contents;
 
   try {
     tags_file_contents = ReadUtf8File( path_to_tag_file );
   } catch ( ... ) {
     return filetype_identifier_map;
```
Comment (fsimonis) [inline cpp/ycm/IdentifierUtils.cpp]: You should be able to write `return {};` (C++11) to default construct the result.
This allows you to move the line `FiletypeIdentifierMap filetype_identifier_map;` further down in front of the loop.

File: cpp/ycm/IdentifierUtils.cpp [hunk 5]
```diff
@@ -114,38 +113,34 @@ const char *const NOT_FOUND = "YCMFOOBAR_NOT_FOUND";
...
-    fs::path path( matches[ 2 ].str() );
-    path = fs::absolute( path, path_to_tag_file.parent_path() )
-           .make_preferred();
-
-    filetype_identifier_map[ filetype ][ path.string() ].push_back( identifier );
+  std::smatch matches;
+  const std::regex expression( TAG_REGEX );
+
+  for ( auto line : tags_file_contents ) {
+    if ( std::regex_search( line, matches, expression ) ) {
```
Comment (fsimonis) [inline cpp/ycm/IdentifierUtils.cpp]: You can define all the `std::string`s in this block as `const`.

File: cpp/ycm/IdentifierUtils.cpp [hunk 6]
```diff
@@ -114,38 +113,34 @@ const char *const NOT_FOUND = "YCMFOOBAR_NOT_FOUND";
...
+      if ( filetype == NOT_FOUND )
+        continue;
+
+      std::string identifier( matches[ 1 ] );
+      fs::path path( matches[ 2 ].str() );
+      path = fs::absolute( path, path_to_tag_file.parent_path() )
+              .make_preferred();
+
+      filetype_identifier_map[ filetype ][ path.string() ]
+              .push_back( identifier );
```
Comment (fsimonis) [inline cpp/ycm/IdentifierUtils.cpp]: You can use `.emplace_back(std::move(identifier));` as it is the last usage of `identifier`.

**Comment by bstaletic** [comment]
I ran the same benchmark with the changes @fsimonis proposed. They made no difference to the performance of `std::regex`. The only change that I noticed in performance is that 1st run of optimised `std::regex` was now stable at 0.6 seconds instead of varying by plus or minus 0.1 second.

I guess the compiler is already smart enough to incorporate these changes instead of the programmer. Either way the regex itself looks farmore readable after the changes and the rest of the changes seem more than reasonable.

> A one time construction would be ideal of course, especially as TAG_REGEXis a const.

The regex is already constructed one time only and is declared `const`. Am I missing anything?

**Comment by bstaletic** [comment]
Now that #723 is merged, this should pass the tests. Is there a way to rerun the tests?

**Comment by vheon** [comment]
@bstaletic I see you are a `Contributor` don't you have access to Travis? I'll do it without problem but just to know if you know how to do it 😃

**Comment by vheon** [comment]
Anyway if you have access to Travis on the right there is a button saying `Restart build`

**Comment by vheon** [comment]
Ops I think I confused `Contributor` with `Collaborator` 😝  sorry, I'll do it.

**Comment by bstaletic** [comment]
The restarted build didn't pick up on using gcc-4.9 and failed. I pushed the commit with changes @fsimonis proposed and it solved that problem as well.

This should be ready for merge now.

@vheon I've seen the `Restart build` on travis on my fork of ycmd, but can't see it here.

**Comment by Valloric** [comment]
:lgtm: 

Thanks for the PR!

---

Review status: 1 of 3 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/719#-:-KhPKiNRU58MZ97esZd2:b-y8s7n3)*

**Comment by bstaletic** [comment]
Can anyone explain the unexpected coverage changes? I won't mind writing some tests, but I can't make sense of codecov's report.

**Comment by Valloric** [comment]
> Can anyone explain the unexpected coverage changes? I won't mind writing some tests, but I can't make sense of codecov's report.

Good question. Codecov provides a "changes", "patch" and "project" coverage in the github summary and I'm not sure I understand them all. I think the only part that is ultimately relevant is this one:

> codecov/project — 93.32% (+0.22%) compared to 13da3d4

Which says that coverage goes up with your PR, which is all I care about.

Don't forget, you can get coverage info locally; you don't have to rely on understanding codecov's UI. 

What I think codecov is trying to say here (with its terrible UX) is that some other lines in the codebase are now not being covered. Might be worth figuring out why that is so, but honestly I wouldn't worry too much about it.

**Comment by vheon** [comment]
The "patch" coverage should be how much of the code of the PR is covered.

**Comment by bstaletic** [comment]
And the "changes" reports if there are "unexpected changes". For this one the documentation only mentions missing tests i.e. test was there but now isn't.

Codecov complains about two files:

- `Utils.cpp` - line: `while( getline( file, line ) )`
- `IdentifierUtils.cpp` - line: `if ( std::regex_search( line, matches, expression ) ) {`

Both lines are partially covered and I'm not sure how to cover them completely.

---

Review status: 1 of 3 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/719#-:-KhTrWkpfdZZfP2spGbN:b88rtug)*

**Comment by puremourning** [comment]
Also worth noting that the codecov guys on Gitter are really good peeps and usually happy to answer reasonable questions :)

**Comment by vheon** [comment]
Just a minor comment otherwise :lgtm:

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cpp/ycm/IdentifierUtils.cpp, line 131 at r7](https://reviewable.io:443/reviews/valloric/ycmd/719#-KhwsYFtpkZEkJ9xnizX:-KhwsYFulESef0nwIO6m:b93lsg1) ([raw file](https://github.com/valloric/ycmd/blob/a8e649b067a0c0f395488e97eab7729bb7f0afb6/cpp/ycm/IdentifierUtils.cpp#L131)):*
> ```cpp
>       const std::string language( matches[ 3 ] );
>       const std::string filetype = FindWithDefault( LANG_TO_FILETYPE,
>                                               language.c_str(),
> ```

Did we messed up the indentation here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/719#-:-Khwsois7xDgyxbHbRAO:bh86k4y)*

**Comment by bstaletic** [comment]
I just heard from Steve from codecov. He asked me to confirm if I want him to review the pull request's coverage data.
Let's see what the latest push reports. If it still reports unexpected changes I'll confirm.

---

Review status: 2 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[cpp/ycm/IdentifierUtils.cpp, line 131 at r7](https://reviewable.io:443/reviews/valloric/ycmd/719#-KhwsYFtpkZEkJ9xnizX:-KhxV0_x1BGExyENYBWv:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/a8e649b067a0c0f395488e97eab7729bb7f0afb6/cpp/ycm/IdentifierUtils.cpp#L131)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

Did we messed up the indentation here?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/719#-:-KhxV5HY0T3jkjq6bWsE:bn5vhxy)*

**Comment by bstaletic** [comment]
Can anyone restart the builds? The errors are spurious.
Travis couldn't download something when running some `apt-get` command.
Appveyor on python2.7 couldn't find CXX compiler.

**Comment by puremourning** [comment]
I kicked tragic but I don't think i can do appveyor

**Comment by Valloric** [comment]
Kicked appveyor.

**Comment by bstaletic** [comment]
I've looked carefully at all the files codecov reports changes on and tried to find an explanation as to why there are changes at all.

- `IdentifierCompleter.cpp` - Specific constructor is now covered, but I have no actual explanation.
- `IdentifierDatabase.cpp` - Range based for statements used to be partially covered. This seems like a nice side effect of using a newer gcc.
- `IdentifierUtils.cpp` - It has one line less to be covered so its relative coverage went down (that one line was covered), but it still only has one partially covered line - only a different one.
- `Result.cpp` - Another partially covered line fully covered. Assuming GCC version.
- `PythonSupport.cpp` - No explanation.
- `ClangUtils.cpp` - Functions that are never used are reported as covered.
- `CompletionData.cpp` - No explanation.
- `Diagnostic.h` - Already was at 0% coverage. No explanation for the diff.
- `TranslationUnitStore.cpp` - No explanation.

In the mean time I asked Steve from codecov to review the PR.

**Comment by bstaletic** [comment]
Oh, forgot about something.

- `utils.py` - Some comments and docstrings are reported as covered. This makes no sense.

**Comment by vheon** [comment]
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/719)*

**Comment by vheon** [comment]
@bstaletic I think that we can safely merge this, if codecov want to check if there are some problem with their product they have all the data here with this PR

**Comment by bstaletic** [comment]
@vheon I'm okay with merging if you guys are. I just don't like seeing failing checks in PR's. On the other hand, I see nothing wrong with the actual code.

**Comment by Valloric** [comment]
Yep, let's merge this. Thanks again!

@homu r+

**Comment by bstaletic** [comment]
> Thanks again!

Don't mention it!

Now we only need to wait for boost to release version 1.64, which should happen today (GMT+0 though). After that I'll post the PR updating the `BoostParts`, `update_boost.py` and `CMakeLists.txt`.
